### PR TITLE
Account balance table in caching layer

### LINKED ISSUE #1024: Account balance table in caching layer. Research + Implement.
Research and define how to track balance of users in our database. Consult plan with @taobun to make sure it works in all cases. Some ideas:
- Track transfer event
- Check if messages or transactions involve what users and reload their balance

Things to be careful of
- Genesis account
- Transaction fees
- etc....

### PR DESCRIPTION
**Pull request by taobun**
fixed #1024

### CODE REVIEW

File: chain/db/account.go [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package db
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+func (b *BandDB) AddOrCreateAccount(address sdk.AccAddress, balance sdk.Coins) error {
```
Comment (sorawit) [inline chain/db/account.go]: Should we calll this `CreateOrUpdateAccount`? Or `SetAccountBalance`?

File: chain/db/db.go [hunk 2]
```diff
@@ -33,21 +35,42 @@ func NewDB(dialect, path string, metadata map[string]string) (*BandDB, error) {
...
+	)
+
+	db.Model(&Transaction{}).AddForeignKey(
+		"block_height",
+		"blocks(height)",
+		"RESTRICT",
+		"RESTRICT",
+	)
+
+	db.Model(&Transaction{}).AddForeignKey(
```
Comment (sorawit) [inline chain/db/db.go:72]: Is it possible that we add transaction before account and violate fk constraint?

File: chain/db/db.go [hunk 3]
```diff
@@ -349,3 +358,45 @@ func (b *BandDB) HandleMessage(txHash []byte, msg sdk.Msg, events map[string]str
 
 	return jsonMap, nil
 }
+
+func (b *BandDB) GetInvolvedAccounts(tx auth.StdTx) []sdk.AccAddress {
+	involvedAccounts := make([]sdk.AccAddress, 0)
+	for _, msg := range tx.GetMsgs() {
+		switch msg := msg.(type) {
+		case zoracle.MsgRequestData:
+			involvedAccounts = append(involvedAccounts, msg.Sender)
```
Comment (sorawit) [inline chain/db/db.go:389]: All data source owners will get fees, right?

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/db/db.go [hunk 4]
```diff
@@ -349,3 +358,45 @@ func (b *BandDB) HandleMessage(txHash []byte, msg sdk.Msg, events map[string]str
...
+			involvedAccounts = append(involvedAccounts, msg.DelegatorAddress)
+		case dist.MsgWithdrawDelegatorReward:
+			involvedAccounts = append(involvedAccounts, msg.DelegatorAddress)
+		case dist.MsgWithdrawValidatorCommission:
+			involvedAccounts = append(involvedAccounts, sdk.AccAddress(msg.ValidatorAddress))
+		case gov.MsgDeposit:
+			involvedAccounts = append(involvedAccounts, msg.Depositor)
+		case gov.MsgSubmitProposal:
+			involvedAccounts = append(involvedAccounts, msg.Proposer)
+		default:
```
Comment (sorawit) [inline chain/db/db.go:427]: Now that we handle a lot of messages. Let's make sure we have case for all messages, and panic for unknown messages, so we fail fast.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/db/table.go [hunk 5]
```diff
@@ -2,18 +2,35 @@ package db
...
+	GasFee      string          `gorm:"not null"`
+	Sender      string          `gorm:"not null"`
+	Success     bool            `gorm:"not null"`
+	BlockHeight int64           `gorm:"not null"`
+	Messages    json.RawMessage `sql:"json;not null"`
+}
+
+type Account struct {
+	Address string `gorm:"primary_key"`
+	Balance string `gorm:"not null"`
```
Comment (sorawit) [inline chain/db/table.go]: Let's store UpdatedHeight here. I think it will help us debug a lot in the future.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/db/db.go (hunk 2 above)
Comment (taobun) [reply chain/db/db.go:72]: A sender is needed to declare before in chain (Cannot send a transaction without account number and sequence number)

**Review by sorawit — DISMISSED** [review]

**Review by Benzbeeb — DISMISSED** [review]

**Review by sorawit — APPROVED** [review]
Let's ship it

**Review by Benzbeeb — APPROVED** [review]

**Pull request merged**